neutron/neutron.go

```diff
@@ -0,0 +1,381 @@
+	requestData := goosehttp.RequestData{ExpectedStatus: []int{http.StatusNoContent}}
+	err := c.client.SendRequest(client.DELETE, "network", "v2.0", url, &requestData)
+	if err != nil {
+		err = errors.Newf(err, "failed to delete floating ip %s details", ipId)
+	}
+	return err
+}
+
+// SecurityGroupRuleV2 describes a rule of a security group. There are 2
+// basic rule types: ingress and egress rules (see RuleInfo struct).
+type SecurityGroupRuleV2 struct {
+	PortRangeMax   *int    `json:"port_range_max"` // Can be nil
+	PortRangeMin   *int    `json:"port_range_min"` // Can be nil
+	IPProtocol     *string `json:"protocol"`       // Can be nil, must be defined if PortRange is used
+	ParentGroupId  string  `json:"security_group_id"`
+	RemoteIPPrefix string  `json:"remote_ip_prefix"`
+	RemoteGroupID  string  `json:"remote_group_id"`
+	EthernetType   string  `json:"ethertype"`
+	Direction      string  `json:"direction"` // Required
+	Id             string  `json:",omitempty"`
+	TenantId       string  `json:"tenant_id,omitempty"`
+}
+
+// SecurityGroupV2 describes a single security group in OpenStack.
+type SecurityGroupV2 struct {
+	Rules       []SecurityGroupRuleV2 `json:"security_group_rules"`
+	TenantId    string                `json:"tenant_id"`
+	Id          string
+	Name        string
+	Description string
+}
+
+// ListSecurityGroupsV2 lists IDs, names, and other details for all security groups.
+func (c *Client) ListSecurityGroupsV2() ([]SecurityGroupV2, error) {
+	var resp struct {
+		Groups []SecurityGroupV2 `json:"security_groups"`
+	}
+	requestData := goosehttp.RequestData{RespValue: &resp}
+	err := c.client.SendRequest(client.GET, "network", "v2.0", ApiSecurityGroupsV2, &requestData)
+	if err != nil {
+		return nil, errors.Newf(err, "failed to list security groups")
+	}
+	return resp.Groups, nil
+}
+
+// SecurityGroupByNameV2 returns the named security group.
+// OpenStack now supports filtering with API calls.
+// More than one Security Group may be returned, as names are not unique
+// e.g. name=default
+func (c *Client) SecurityGroupByNameV2(name string) ([]SecurityGroupV2, error) {
+	var resp struct {
+		Groups []SecurityGroupV2 `json:"security_groups"`
+	}
+	// Can't have any spaces in a url
+	noSpaceName := name
+	if strings.Contains(name, " ") {
+		noSpaceName = strings.Replace(noSpaceName, " ", "+", -1)
+	}
+	url := fmt.Sprintf("%s?name=%s", ApiSecurityGroupsV2, noSpaceName)
+	requestData := goosehttp.RequestData{RespValue: &resp}
+	err := c.client.SendRequest(client.GET, "network", "v2.0", url, &requestData)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Groups) == 0 {
+		return nil, errors.Newf(err, "failed to find security group with name: %s", name)
+	}
+	return resp.Groups, nil
+}
+
+// CreateSecurityGroupV2 creates a new security group.
+func (c *Client) CreateSecurityGroupV2(name, description string) (*SecurityGroupV2, error) {
+	var req struct {
+		SecurityGroupV2 struct {
```

Comment:
A rules array isn't optional when a SecurityGroup is returned from neutron, trying to stay consistent with the API.